indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedResourceManagementStrategy.java

```diff
@@ -0,0 +1,360 @@
+          workerConfig.getAutoScaler().ipToIdLookup(
+              Lists.newArrayList(
+                  Iterables.transform(
+                      runner.getLazyWorkers(),
+                      new Function<ZkWorker, String>()
+                      {
+                        @Override
+                        public String apply(ZkWorker input)
+                        {
+                          return input.getWorker().getIp();
+                        }
+                      }
+                  )
+              )
+          )
+      );
+
+      final Set<String> stillExisting = Sets.newHashSet();
+      for (String s : currentlyTerminating) {
+        if (workerNodeIds.contains(s)) {
+          stillExisting.add(s);
+        }
+      }
+      currentlyTerminating.clear();
+      currentlyTerminating.addAll(stillExisting);
+
+      if (currentlyTerminating.isEmpty()) {
+        final int scaleDownTarget = getScaleDownTarget(zkWorkers, workerConfig);
+        final Predicate<ZkWorker> isLazyWorker = ResourceManagementUtil.createLazyWorkerPredicate(config);
+        final List<String> laziestWorkerIps =
+            Lists.transform(
+                runner.markWorkersLazy(isLazyWorker, scaleDownTarget),
+                new Function<ZkWorker, String>()
+                {
+                  @Override
+                  public String apply(ZkWorker zkWorker)
+                  {
+                    return zkWorker.getWorker().getIp();
+                  }
+                }
+            );
+        if (laziestWorkerIps.isEmpty()) {
+          log.debug("Found no lazy workers");
+        } else {
+          log.info(
+              "Terminating %,d lazy workers: %s",
+              laziestWorkerIps.size(),
+              Joiner.on(", ").join(laziestWorkerIps)
+          );
+
+          final AutoScalingData terminated = workerConfig.getAutoScaler().terminate(laziestWorkerIps);
+          if (terminated != null) {
+            currentlyTerminating.addAll(terminated.getNodeIds());
+            lastTerminateTime = new DateTime();
+            scalingStats.addTerminateEvent(terminated);
+            didTerminate = true;
+          }
+        }
+      } else {
+        Duration durSinceLastTerminate = new Duration(lastTerminateTime, new DateTime());
+
+        log.info("%s terminating. Current wait time: %s", currentlyTerminating, durSinceLastTerminate);
+
+        if (durSinceLastTerminate.isLongerThan(config.getMaxScalingDuration().toStandardDuration())) {
+          log.makeAlert("Worker node termination taking too long!")
+             .addData("millisSinceLastTerminate", durSinceLastTerminate.getMillis())
+             .addData("terminatingCount", currentlyTerminating.size())
+             .emit();
+
+          currentlyTerminating.clear();
+        }
+      }
+
+      return didTerminate;
+    }
+  }
+
+  private int getScaleDownTarget(Collection<ZkWorker> zkWorkers, WorkerBehaviorConfig workerConfig)
```

Comment:
maxWorkersToTerminate might be more descriptive of what this actually returns? target could be misunderstood as the number of remaining workers after termination.